### DISCUSSION TITLE
Regarding Quantifiers Not working for String generation

### DISCUSSION THREAD

**Discussion by AnkitMU** [cvc5/cvc5#10207] [General]
I'm trying to generate the String value Using Universal quantifiers, then checking the (check-sat) fot the model.
here the results is comimng is unknow,
but without using the quantifiers , it is generating the output.

The codeis mnetioned below.
(set-logic ALL)
(set-option :produce-models true)
(set-option :strings-exp true)

(declare-fun first_name (Int) String)
(declare-fun last_name (Int) String)
(declare-fun email (Int) String)
(declare-fun address (Int) String)
(declare-fun Account_number (Int) Int)
(declare-fun Person (Int) Bool)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (=(str.len (first_name o)) 5)
            (str.in_re (first_name o) (re.* (re.range "a" "z")))
        )
    ))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (=(str.len (last_name o)) 5)
            (str.in_re (last_name o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))
        )
    ))
)

(assert
    (forall ((o Int)) (=> (Person o)
        (not (= (first_name o) (last_name o)))
    ))
)

(assert
    (forall ((o Int)) (=> (Person o)
        (= (email o) (str.++ (first_name o) "." (last_name o) "@mu.ie"))
    ))
)

(assert
    (forall ((o Int)) (=> (Person o)
        (str.suffixof "mu.ie" (email o))
    ))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (> (str.len (address o)) 14)
            (str.in_re (address o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))
        )
    ))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (<= (Account_number o) 9999999999)
            (>= (Account_number o) 1000000000)
        )
    ))
)

(assert (exists ((o Int)) (Person o)))

(check-sat)
(get-value ((first_name 1)(last_name 1)(email 1)(address 1)(Account_number 1)))

The generated output is
**unknown**
(((first_name 1) "aaaaa") ((last_name 1) "AAAAA") ((email 1) "aaaaa.AAAAA@mu.ie") ((address 1) "AAAAAAAAAAAAAAA") ((Account_number 1) 1000000000))

**Comment by ajreynol** [comment]
By default, cvc5 will typically not answer "sat" in the presence of quantified formulas.

To enable techniques to finding models with quantified formulas, you can try a number of options including `--mbqi`, `--finite-model-find`, `--sygus-inst`.

For the above example,`--mbqi` is able to find "sat" quickly.

**Comment by AnkitMU** [comment]
I'm trying to generate the String value in CVC5. But each time im running this model in cvc5 Im getting the same **Output**(please refer below) , although im using **distinct** function. but its giving me same output and there No diversity in the result, to generate First name im using the constraint as   (str.in_re (first_name o) (re.* (re.range "a" "z"))) but it is returning me "aaaaaaaaaaaaaaa". For last name  (str.in_re (last_name o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))  but it is returning me "AAAAAAAAAAAAAAA".
Am I doing something wrong in the below SMTlib2 formula. How Can I solve this problem. (i) get different result on each model produce (ii) get a more diverse Output as stated in constraint.

(set-logic ALL)
(set-option :produce-models true)
(set-option :strings-exp true)

(declare-fun first_name (Int) String)
(declare-fun last_name (Int) String)
(declare-fun email (Int) String)
(declare-fun address (Int) String)
(declare-fun Account_number (Int) Int)

(define-fun Person ((o Int)) Bool
    (and (>= o 1) (<= o 10))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (> (str.len (first_name o)) 5)
            (str.in_re (first_name o) (re.* (re.range "a" "z")))
        )
    ))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (> (str.len (last_name o)) 5)
            (str.in_re (last_name o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))
        )
    ))
)

(assert
    (forall ((o Int)) (=> (Person o)
        (not (= (first_name o) (last_name o)))
    ))
)

(assert
    (forall ((o Int)) (=> (Person o)
        (= (email o) (str.++ (first_name o) "." (last_name o) "@mu.ie"))
    ))
)

(assert
    (forall ((o Int)) (=> (Person o)
        (str.suffixof "mu.ie" (email o))
    ))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (> (str.len (address o)) 14)
            (str.in_re (address o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))
        )
    ))
)

(assert 
    (forall ((o Int)) (=> (Person o)
        (and
            (<= (Account_number o) 9999999999)
            (>= (Account_number o) 1000000000)
        )
    ))
)

(assert (exists ((o Int)) (Person o)))

(assert (distinct 
(str.++ (first_name 1) "1")
(str.++ (first_name 2) "2")
(str.++ (first_name 3) "3")
(str.++ (last_name 1) "1")
(str.++ (last_name 2) "2")
(str.++ (last_name 3) "3")
(str.++ (address 1) "1")
(str.++ (address 2) "2")
(str.++ (address 3) "3")
))
(assert (distinct 
    (Account_number 1)
    (Account_number 2)
    (Account_number 3)
))

(check-sat)
(get-value ((first_name 1)(last_name 1)(email 1)(address 1)(Account_number 1)))
(get-value ((first_name 2)(last_name 2)(email 2)(address 2)(Account_number 2)))
(get-value ((first_name 3)(last_name 3)(email 3)(address 3)(Account_number 3)))


**OUTPUT**
sat
(((first_name 1) "aaaaaaaaaaaaaaa") ((last_name 1) "AAAAAAAAAAAAAAA") ((email 1) "aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie") ((address 1) "ZAAAAAAAAAAAAAA") ((Account_number 1) 1000000000))
(((first_name 2) "aaaaaaaaaaaaaaa") ((last_name 2) "AAAAAAAAAAAAAAA") ((email 2) "aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie") ((address 2) "ZAAAAAAAAAAAAAA") ((Account_number 2) 1000000001))
(((first_name 3) "aaaaaaaaaaaaaaa") ((last_name 3) "AAAAAAAAAAAAAAA") ((email 3) "aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie") ((address 3) "ZAAAAAAAAAAAAAA") ((Account_number 3) 1000000002))
